### PR TITLE
Introduce Graph benchmarking foundations

### PR DESCRIPTION
**Pull request by Alfred-Mountfield**
## 🌟 What is the purpose of this PR?

We are likely to engage in some early optimisation work soon. It would be good to start tracking performance before we do so, in order to objectively reason about the improvements we're making.

This PR introduces the foundational boilerplate for benchmarking _the Graph Query Layer_. It utilises [Criterion.rs](https://bheisler.github.io/criterion.rs/book/criterion_rs.html), and is inspired by the [Diesel Benchmarks](https://github.com/diesel-rs/diesel/blob/master/diesel_bench).

## 🔗 Related links

- [Asana task](https://app.asana.com/0/0/1203065722948213/f) _(internal)_

## 🔍 What does this change?

- Introduces `bench` that sits alongside `bin`, `lib`, and `tests`
- Adds the basic scaffolding for `criterion` benchmarks
- Adds setup code to create a new database for a suite of benchmarks, including teardown logic
- Adds demo code to seed the database for a suite of read benchmarks
- Adds a demo benchmark to test the times required to query a specific entity in the datastore

## 📜 Does this require a change to the docs?

- The README will need updating to say how to run the Benchmarks. This should be done when the Makefile.toml is updated

## ⚠️ Known issues

- The Postgres Superuser credentials are currently not always easily accessible to the graph. Depending on the environment we run these benchmarks in we'll need to think about if we put them in the environment. At the moment they are hardcoded to the defaults
- The benchmarks currently copy the main dev database. If the current database has been polluted via integration test runs, or similar, then **this state will pollute the benchmarks**. It is possible to write a query to truncate all of the tables, however at the moment I believe copying the state is actually beneficial. It means that people can run the benchmarks within their current databases without creating new synthetic data (should they wish), and it gives us more flexibility later on to choose how to seed the various states of the graph.
- There is a non-insignificant amount of duplication between the Integration Tests set up and the Benchmarks. We probably can create a unified `test-util` crate or something to hold a lot of this common code. It'd be good if we could even re-use a lot of the sample graphs and structural queries.

## 🐾 Next steps

- Introduce 'representative' benchmarks for all of the major basic read operations
- Introduce 'representative' benchmarks for all of the major basic write operations
- Introduce benchmarks to analyse how the major read operations behave at varying scales of data in the graph
- Update the Makefile to run the benchmarks
- Run the benchmarks as part of the CI
  - Persist the results of the benchmarks in such a way that we can track progress 
- Figure out a way to create a synthetic 'representative' test graph
- Improve the mechanism through which we express states of the graph, especially useful for tests and benchmarks

## 🛡 What tests cover this?

- New ones

## ❓ How to test this?

1. Checkout the branch
2. Run `cargo make deployment-up`
3. Run `make recreate-db`
4. Run `make migrate-up`
5. Run `cargo bench`

## 📹 Demo

image

### CODE REVIEW

File: packages/graph/hash_graph/bench/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+serde = { version = "1.0.137", features = ["derive"] }
+serde_json = "1.0.83"
+tokio = { version = "1.18.2", features = ["rt-multi-thread", "macros"] }
+tokio-postgres = { version = "0.7.6", default-features = false }
+type-system = { git = "https://github.com/blockprotocol/blockprotocol", rev = "6526c0e" }
+uuid = { version = "1.1.2", features = ["v4", "serde"] }
+
+[[bench]]
+# TODO - Rename, lib.rs, main.rs, hash_graph_benchmark.rs?
+name = "entry"
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/bench/Cargo.toml:29]: I couldn't come up with a good name for this. As expressed in the comment above `autobenches`, we have a single entrypoint to the benchmark 'binary' (even though we may have many benchmarks). 

This kind of makes me think `main.rs` won't be too misleading?

File: packages/graph/hash_graph/bench/benches/representative_read/mod.rs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    store::{AccountStore, EntityStore},
+};
+use graph_test_data::{data_type, entity, entity_type, link_type, property_type};
+use rand::{seq::IteratorRandom, thread_rng};
+use type_system::uri::{BaseUri, VersionedUri};
+use uuid::Uuid;
+
+use crate::util::{setup, StoreWrapper};
+
+/// DOC - TODO
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/bench/benches/representative_read/mod.rs:20]: Not sure this is worth an Asana task, it can be address as the `Samples` object is fleshed out and the shape of it is better defined

File: packages/graph/hash_graph/bench/benches/representative_read/mod.rs [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    pub entities: HashMap<AccountId, HashMap<VersionedUri, Vec<EntityId>>>,
+}
+
+/// Sets up the sample "representative" environment in which operations are benchmarked.
+///
+/// This initializes the database for all benchmarks within this module, and therefore should be a
+/// single point to swap out the seeding of test data when we can invest time in creating a
+/// representative environment.
+async fn seed_db(store_wrapper: &mut StoreWrapper) -> Samples {
+    // TODO: We'll want to test distribution across accounts
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/bench/benches/representative_read/mod.rs:31]: I'll make and attach an asana task if people agree this is worth doing at some point (probably when we have sharding in place)

File: packages/graph/hash_graph/bench/benches/util.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    },
+};
+use tokio::runtime::Runtime;
+use tokio_postgres::NoTls;
+use type_system::{DataType, EntityType, LinkType, PropertyType};
+
+type Pool = PostgresStorePool<NoTls>;
+pub type Store = <Pool as StorePool>::Store<'static>;
+
+// TODO - deduplicate with integration/postgres/mod.rs
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/bench/benches/util.rs:17]: This was mentioned in the PR description, will create and attach an asana task if people agree with it

File: packages/graph/hash_graph/bench/benches/util.rs [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+    bench_db_name: String,
+    source_db_pool: Pool,
+    _pool: ManuallyDrop<Pool>,
+    pub store: ManuallyDrop<Store>,
+}
+
+impl StoreWrapper {
+    pub async fn new(bench_db_name: &str) -> Self {
+        let source_db_connection_info = DatabaseConnectionInfo::new(
+            // TODO - get these from env
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/bench/benches/util.rs:28]: This was also mentioned in the PR description, will create and attach an asana task if people agree with it

File: packages/graph/hash_graph/tests/integration/postgres/mod.rs [hunk 6]
```diff
@@ -112,7 +111,7 @@ impl DatabaseTestWrapper {
         for property_type in property_types {
             store
                 .create_property_type(
-                    PropertyType::from_str(property_type).expect("could not parse data type"),
+                    PropertyType::from_str(property_type).expect("could not parse property type"),
```
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/tests/integration/postgres/mod.rs:114]: Drive-by

File: packages/graph/hash_graph/bench/benches/representative_read/mod.rs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    let (runtime, mut store_wrapper) = setup("representative_read");
+
+    let samples = runtime.block_on(async { seed_db(&mut store_wrapper).await });
+    let store = &store_wrapper.store;
+
+    for (account_id, type_ids_and_entity_ids) in samples.entities {
+        for (entity_type_id, entity_ids) in type_ids_and_entity_ids {
+            group.bench_with_input(
+                BenchmarkId::from_parameter(format!(
+                    "Account ID: `{}`, Entity Type ID: `{}`",
```
Comment (thehabbos007) [inline packages/graph/hash_graph/bench/benches/representative_read/mod.rs:105]: The parameter is being truncated in the terminal output, i wonder if we want to change this format to make it easier to understand the output. For example I have an output like:
```
Benchmarking representative_read/Account ID: `d4e16033-c281-4cde-aa35-9085bf2e7579`, Entity Type ID: `https://blo...: 
Benchmarking representative_read/Account ID: `d4e16033-c281-4cde-aa35-9085bf2e7579`, Entity Type ID: `https://blo...: 
representative_read/Account ID: `d4e16033-c281-4cde-aa35-9085bf2e7579`, Entity Type ID: `https://blo...
```
which isn't really that helpful. I can imagine the actual result files containing the non-truncated versions, but maybe there is a better way to attach these params?

File: packages/graph/hash_graph/bench/benches/representative_read/mod.rs [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        1,
+    );
+
+    let store = &mut store_wrapper.store;
+
+    let mut entity_ids = Vec::with_capacity(100);
+    for _ in 0..100 {
+        entity_ids.push(
+            store
+                .create_entity(person.clone(), person_id.clone(), account_id, None)
```
Comment (thehabbos007) [inline packages/graph/hash_graph/bench/benches/representative_read/mod.rs:72]: I looked into if postgres dedupliated JSONB data or not, and it seems that it does not. Just flagging as it means our testing strategy might be fine. If it did deduplicate entries, we'd have an unrepresentative DB :+1: 
https://www.postgresql.org/docs/current/btree-implementation.html#BTREE-DEDUPLICATION

**Review by thehabbos007 — COMMENTED** [review]
I have some comments here, some are answering your comments :)

File: packages/graph/hash_graph/bench/Cargo.toml (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/Cargo.toml:29]: One of my main aversions is I don't like `bench/benches/bench.rs` 😂 Not fully against it though

File: packages/graph/hash_graph/bench/benches/util.rs [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        {
+            let conn = source_db_pool
+                .acquire_owned()
+                .await
+                .expect("could not acquire a database connection");
+            let client = conn.as_client();
+
+            let _ = client.execute(
+                r#"
+                        /* KILL ALL EXISTING CONNECTION FROM ORIGINAL DB (sourcedb)*/
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/util.rs:51]: Yes, at least [according to StackOverflow](https://stackoverflow.com/a/876565/14687716) this was true a long time ago

[the remark this answers is not in the document]
Comment (thehabbos007) [reply packages/graph/hash_graph/bench/benches/util.rs:51]: Seems that the docs specify this as well: 

It is possible to create additional template databases, and indeed one can copy any database in a cluster by specifying its name as the template for CREATE DATABASE. It is important to understand, however, that this is not (yet) intended as a general-purpose “COPY DATABASE” facility. The principal limitation is that no other sessions can be connected to the source database while it is being copied. CREATE DATABASE will fail if any other connection exists when it starts; during the copy operation, new connections to the source database are prevented.
https://www.postgresql.org/docs/current/manage-ag-templatedbs.html

File: packages/graph/hash_graph/bench/benches/representative_read/mod.rs (hunk 8 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/representative_read/mod.rs:72]: Oh to be clear I do not think this database is representative at all right now 😁 See next steps:
> Figure out a way to create a synthetic 'representative' test graph

File: packages/graph/hash_graph/bench/benches/util.rs [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                        /* CLONE DATABASE TO NEW ONE(TARGET_DB) */
+                        CREATE DATABASE {bench_db_name} WITH TEMPLATE {} OWNER {};
+                        "#,
+                        source_db_connection_info.database(),
+                        source_db_connection_info.user()
+                    ),
+                    &[],
+                )
+                .await
+                .expect("failed to clone database");
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/bench/benches/util.rs:71]:
```suggestion
            // CLONE DATABASE TO NEW ONE(TARGET_DB)
            client
                .execute(
                    &format!(
                        r#"
                        CREATE DATABASE {bench_db_name} WITH TEMPLATE {} OWNER {};
                        "#,
                        source_db_connection_info.database(),
                        source_db_connection_info.user()
                    ),
                    &[],
                )
                .await
                .expect("failed to clone database");
```

File: packages/graph/hash_graph/bench/benches/representative_read/knowledge/entity.rs [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+    store: &Store,
+    entity_ids: &[EntityId],
+) {
+    b.to_async(runtime).iter_batched(
+        || {
+            // Each iteration, *before timing*, pick a random entity from the sample to query
+            *entity_ids.iter().choose(&mut thread_rng()).unwrap()
+        },
+        |entity_id| async move {
+            let _ = store
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/representative_read/knowledge/entity.rs:37]: CLion does not like that 😬

File: packages/graph/hash_graph/bench/benches/util.rs [hunk 12]
```diff
@@ -0,0 +1,195 @@
...
+                .await
+                .expect("failed to create entity type");
+        }
+    }
+}
+
+impl Drop for StoreWrapper {
+    fn drop(&mut self) {
+        // SAFETY: We're in the process of dropping the parent struct, we just need to ensure we
+        //  release the connections of this pool before deleting the database
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/bench/benches/util.rs:160]: Appreciate you have written a comment, about what we do in the method, but this does not satisfy a SAFETY. The section should tell, why we are allowed to use `unsafe` here, so it remains safe. This is typically written in the `# Safety` section on a method. In this case, it's not allowed to use `.store` or `.pool` after it's dropped
```suggestion
        // We're in the process of dropping the parent struct, we just need to ensure we release
        // the connections of this pool before deleting the database
        // SAFETY: The values of `store` and `pool` are not accessed after dropping
 
```

File: packages/graph/hash_graph/bench/benches/util.rs [hunk 13]
```diff
@@ -0,0 +1,195 @@
...
+use type_system::{DataType, EntityType, LinkType, PropertyType};
+
+type Pool = PostgresStorePool<NoTls>;
+pub type Store = <Pool as StorePool>::Store<'static>;
+
+// TODO - deduplicate with integration/postgres/mod.rs
+pub struct StoreWrapper {
+    bench_db_name: String,
+    source_db_pool: Pool,
+    _pool: ManuallyDrop<Pool>,
```
Comment (TimDiekmann) [inline packages/graph/hash_graph/bench/benches/util.rs:21]: `_pool` is not unused, can we rename it to `pool`?

**Review by TimDiekmann — CHANGES_REQUESTED** [review]

File: packages/graph/hash_graph/bench/benches/util.rs (hunk 13 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/util.rs:21]: Added

File: packages/graph/hash_graph/bench/benches/util.rs (hunk 12 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/util.rs]: Cheers

File: packages/graph/hash_graph/bench/benches/util.rs (hunk 10 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/util.rs]: Have committed something similar, thanks

File: packages/graph/hash_graph/bench/benches/util.rs [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+            let client = conn.as_client();
+
+            let _ = client.execute(
+                r#"
+                        /* KILL ALL EXISTING CONNECTION FROM ORIGINAL DB (sourcedb)*/
+                        SELECT pg_terminate_backend(pg_stat_activity.pid) FROM pg_stat_activity
+                        WHERE pg_stat_activity.datname = '{DEV_DATABASE}' AND pid <> pg_backend_pid();
+                        "#,
+                &[],
+            ).await.expect("failed to kill existing connections");
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/util.rs]: Have committed something similar, thanks

File: packages/graph/hash_graph/bench/benches/representative_read/mod.rs [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+        entities: HashMap::from([(account_id, HashMap::from([(person_id, sampled_entity_ids)]))]),
+    }
+}
+
+#[criterion]
+fn bench_representative_read(c: &mut Criterion) {
+    let mut group = c.benchmark_group("representative_read");
+    let (runtime, mut store_wrapper) = setup("representative_read");
+
+    let samples = runtime.block_on(async { seed_db(&mut store_wrapper).await });
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/representative_read/mod.rs]: :( thanks

File: packages/graph/hash_graph/bench/benches/util.rs [hunk 16]
```diff
@@ -0,0 +1,195 @@
...
+
+            let _ = client
+                .execute(
+                    &format!(
+                        r#"
+                        /* CLONE DATABASE TO NEW ONE(TARGET_DB) */
+                        CREATE DATABASE {bench_db_name} WITH TEMPLATE {} OWNER {};
+                        "#,
+                        source_db_connection_info.database(),
+                        source_db_connection_info.user()
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/util.rs]: I did think about this when creating it but considering it was a new database I was unsure if we even needed to continue accessing it under the `graph` user. Probably a good idea to do so, was easy enough to swap out now without parameterising by env-vars though. Thanks!

File: packages/graph/hash_graph/bench/benches/representative_read/knowledge/entity.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/graph/hash_graph/bench/benches/representative_read/knowledge/entity.rs:37]: We can remove the whole `let _ = ` though 🙂

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/representative_read/knowledge/entity.rs]: Removed

**Review by TimDiekmann — DISMISSED** [review]
Great to see some progress here, thank you!

**Review by thehabbos007 — DISMISSED** [review]
Looks good to me as well. Thanks Alfie!

File: packages/graph/hash_graph/bench/benches/util.rs (hunk 5 above)
Comment (Alfred-Mountfield) [reply packages/graph/hash_graph/bench/benches/util.rs:28]: This has been done

**Review by thehabbos007 — APPROVED** [review]
:rocket:

**Review by TimDiekmann — APPROVED** [review]
